org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java

```diff
@@ -964,6 +987,185 @@ public void recyclableThreads(ThreadReference thread) {
+                            String.format("SetVariableRequest: Variable %s cannot be found.", arguments.name));
+                }
+            } catch (IllegalArgumentException | AbsentInformationException | InvalidTypeException
+                    | UnsupportedOperationException | ClassNotLoadedException e) {
+                return new Responses.ErrorResponseBody(convertDebuggerMessageToClient(e.getMessage()));
+            }
+            int referenceId = getReferenceId(thread, newValue, showStaticVariables);
+
+            int indexedVariables = 0;
+            if (newValue instanceof ArrayReference) {
+                indexedVariables = ((ArrayReference) newValue).length();
+            }
+            return new Responses.SetVariablesResponseBody(
+                    this.variableFormatter.typeToString(newValue == null ? null : newValue.type(), options), // type
+                    this.variableFormatter.valueToString(newValue, options), // value,
+                    referenceId, indexedVariables);
+
+        }
+
+        private Value setValueProxy(Type type, String value, SetValueFunction setValueFunc, Map<String, Object> options)
+                throws ClassNotLoadedException, InvalidTypeException {
+            IValueFormatter formatter = getFormatterForModification(type, options);
+            Value newValue = formatter.valueOf(value, type, options);
+            setValueFunc.apply(newValue);
+            return newValue;
+        }
+
+        private IValueFormatter getFormatterForModification(Type type, Map<String, Object> options) {
+            char signature0 = type.signature().charAt(0);
+
+            if (signature0 == LONG || signature0 == INT || signature0 == SHORT || signature0 == BYTE || signature0 == FLOAT
+                    || signature0 == DOUBLE || signature0 == BOOLEAN || signature0 == CHAR
+                    || type.signature().equals(STRING_SIGNATURE)) {
+                return this.variableFormatter.getValueFormatter(type, options);
+            }
+            throw new UnsupportedOperationException(String.format("Set value for type %s is not supported.", type.name()));
+        }
+
+        private Value setStaticFieldValue(Type declaringType, Field field, String name, String value, Map<String, Object> options)
+                throws ClassNotLoadedException, InvalidTypeException {
+            if (field.isFinal()) {
+                throw new UnsupportedOperationException(
+                        String.format("SetVariableRequest: Final field %s cannot be changed.", name));
+            }
+            if (!(declaringType instanceof ClassType)) {
+                throw new UnsupportedOperationException(
+                        String.format("SetVariableRequest: Field %s in interface cannot be changed.", name));
+            }
+            return setValueProxy(field.type(), value, newValue -> ((ClassType) declaringType).setValue(field, newValue), options);
```

Comment:
interface SetValueFunction {
        void apply(Value value) throws InvalidTypeException, ClassNotLoadedException;
    }
because SetValueFunction is able to defining the throws clause, so I don't need to catch these exceptions, just to delegate these exceptions to the callee.